rpcs3/Emu/Cell/lv2/sys_net/lv2_socket_p2ps.cpp

```diff
@@ -0,0 +1,800 @@
+
+	lv2_socket_p2ps* sock_client = reinterpret_cast<lv2_socket_p2ps*>(idm::check_unlocked<lv2_socket>(p2ps_client));
+	{
+		std::lock_guard lock(sock_client->mutex);
+		paddr->sin_family = SYS_NET_AF_INET;
+		paddr->sin_addr   = std::bit_cast<be_t<u32>, u32>(sock_client->op_addr);
+		paddr->sin_port   = sock_client->op_vport;
+		paddr->sin_vport  = sock_client->op_port;
+		paddr->sin_len    = sizeof(sys_net_sockaddr_in_p2p);
+	}
+
+	return {true, p2ps_client, ps3_addr};
+}
+
+s32 lv2_socket_p2ps::bind(const sys_net_sockaddr& addr, s32 ps3_id)
+{
+	const auto* psa_in_p2p = reinterpret_cast<const sys_net_sockaddr_in_p2p*>(&addr);
+
+	// For SYS_NET_SOCK_STREAM_P2P sockets, the port is the "fake" tcp port and the vport is the udp port it's bound to
+	u16 p2p_port  = psa_in_p2p->sin_vport;
+	u16 p2p_vport = psa_in_p2p->sin_port;
+
+	sys_net.notice("[P2PS] Trying to bind %s:%d:%d", np::ip_to_string(std::bit_cast<u32>(psa_in_p2p->sin_addr)), p2p_port, p2p_vport);
+
+	ensure(p2p_vport != 0);
+	if (p2p_port != 3658)
+	{
+		sys_net.warning("[P2PS] Attempting to bind a socket to a port != 3658");
+	}
+
+	socket_type real_socket{};
+
+	auto& nc = g_fxo->get<network_context>();
+	{
+		std::lock_guard list_lock(nc.list_p2p_ports_mutex);
+		if (!nc.list_p2p_ports.contains(p2p_port))
+		{
+			nc.list_p2p_ports.emplace(std::piecewise_construct, std::forward_as_tuple(p2p_port), std::forward_as_tuple(p2p_port));
+		}
+
+		auto& pport = nc.list_p2p_ports.at(p2p_port);
+		real_socket = pport.p2p_socket;
+		{
+			// Ensures the socket & the bound list are updated at the same time to avoid races
+			std::lock_guard vport_lock(pport.bound_p2p_vports_mutex);
+			std::lock_guard sock_lock(mutex);
+
+			const u64 key = (static_cast<u64>(p2p_vport) << 32);
+			pport.bound_p2p_streams.emplace(key, ps3_id);
+
+			port   = p2p_port;
+			vport  = p2p_vport;
+			socket = real_socket;
+		}
+	}
+
+	return CELL_OK;
+}
+
+std::optional<s32> lv2_socket_p2ps::connect(const sys_net_sockaddr& addr)
+{
+	std::lock_guard lock(mutex);
+
+	p2ps_encapsulated_tcp send_hdr;
+	const auto psa_in_p2p = reinterpret_cast<const sys_net_sockaddr_in_p2p*>(&addr);
+	auto name             = sys_net_addr_to_native_addr(addr);
+
+	// This is purposefully inverted, not a bug
+	const u16 dst_vport = psa_in_p2p->sin_port;
+	const u16 dst_port  = psa_in_p2p->sin_vport;
+
+	socket_type real_socket{};
+
+	auto& nc = g_fxo->get<network_context>();
+	{
+		std::lock_guard list_lock(nc.list_p2p_ports_mutex);
+		if (!nc.list_p2p_ports.contains(port))
+			nc.list_p2p_ports.emplace(std::piecewise_construct, std::forward_as_tuple(port), std::forward_as_tuple(port));
+
+		auto& pport = nc.list_p2p_ports.at(port);
+		real_socket = pport.p2p_socket;
+		{
+			std::lock_guard lock(pport.bound_p2p_vports_mutex);
+			if (vport == 0)
+			{
+				// Unassigned vport, assigns one
+				sys_net.warning("[P2PS] vport was unassigned before connect!");
+				u16 found_vport = 30000;
+				while (true)
+				{
+					found_vport++;
```

Comment:
so vport 30000 is ignored? just curious

Reply:
Yes, tbh I think I took the range randomly so it has no real meaning(beyond being a "high" range).